Set up entry points and connect them to the scraper.

ElenaKusevska opened the pull request:
@rgaudin This is the code to define the entry points and connect them to the scraper. Please review the code, and if it's OK merge the PR, or let me know if I should fix anything.

ElenaKusevska committed:
create file/directory structure

ElenaKusevska committed:
add entrypoints

ElenaKusevska committed:
connect entrypoints to scraper

ElenaKusevska committed:
rename __main.py__ to __main__.py

ElenaKusevska committed:
set debug and enable logging

ElenaKusevska committed:
more detailed commandline arguments

ElenaKusevska committed:
use commandline arguments in scraper self definition

ElenaKusevska committed:
remove commons parameter. Add options for language

rgaudin commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+requests>=2.26.0,<2.27
+Jinja2>=2.1,<3.0.1
```
You need to add zimscraperlib (that you're using) and libzim as well

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -0,0 +1,26 @@
+
+class wikihow2zim:
+    def __init__(self, **kwargs):
+        print("kwargs", kwargs)
```
this is debug line that we don't merge

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -0,0 +1,26 @@
+            if option not in kwargs:
+                raise ValueError(f"Missing parameter `{option}`")
+
+        self.debug = kwargs.get("debug")
```
For simple assignments, you can factorize this code to save several almost identical lines.

rgaudin commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+        "Programming Language :: Python :: 3.8",
+        "License :: OSI Approved :: GNU General Public License v3 or later (GPLv3+)",
+    ],
+    python_requires=">=3.8",
```
Most our scrapers targets python `3.6+`. Reasoning being that we want them to be ran on the latest Ubuntu LTS and debian stable.
Given this is a new scraper, that Ubuntu 20.04 already has python3.8 and that debian 11 (python 3.9) will become stable this month, I think it's good to start requiring 3.8+ with this scraper.

rgaudin commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+    description="Make ZIM file from WikiHow articles",
+    long_description=read("README.md"),
+    long_description_content_type="text/markdown",
+    author=" ",
```
Please fill-in those metadata. You can use `Kiwix Team` and `dev@kiwix.org` as well as `https://kiwix.org/`

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,66 @@
+
+    parser.add_argument(
+        "--language",
+        choices=['en', 'de', 'fr'],
```
You should define the list of actual domains we'll support in constants and use it here in the choices and help message.

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,66 @@
+def main():
+    parser = argparse.ArgumentParser(
+        prog=NAME,
+        description="Scraper to create ZIM files from Kolibri channels",
```
Forgot to customize this 😀

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,66 @@
+
+    parser.add_argument(
+        "--low-quality",
+        help="Uses only the `low_res` version of videos if available. "
```
Wikihow only offer a single video version. We'll use this version to decide which encode profile to use so `Use lower-quality, smaller file-size video encode` would be better

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,66 @@
+
+    parser.add_argument(
+        "--no-external-links",
+        help="If specified, don't include external links",
```
Don't write `If specified` as it is implied by the use of the param. When writting this on a single param, it may confuse user into thinking there's a special behavior.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -0,0 +1,26 @@
+        self.s3_url_with_credentials = kwargs.get("s3_url_with_credentials")
+
+    def run(self):
+        print("self", self.__dict__)
```
We never print in scrapers. We always use the logger so that we get a timestamp and can filter on level as well.

rgaudin reviewed (changes requested):
Thank you @ElenaKusevska, good start!

I've commented in-line for some adjustments. In addition:

* Use [black](https://github.com/psf/black) formatting. You should ensure you use black before every commit as applying it afterwards messes up with the whole git history (many non-code changes). You can even configure a https://pre-commit.com/hooks.html if you'd like.
* entrypoint lacks all the common options:
    - all the Zim metadata: name, title, description, publisher, tags
    - other usage params: output, tmp_dir, zim-file, keep, version
* Use a better PR name, like `Basic scraper structure` in this case
* Use the “Linked Issues” feature of github to link the issue you are fixing (#2) so that it gets closed when we merge to master

We usually squash the meaningless commits (19ab4b1 for instance) using `git rebase` before or during the PR so that the git history stays informative and useable when looking back at it. It's OK for this PR but keep it in mind for next ones.

Let me know if I can assist you in any way.

ElenaKusevska committed:
Include additional commandline arguments. Remove print statements. Small corrections.

ElenaKusevska committed:
delete __main.py__

ElenaKusevska committed:
add urls for wikihow homepage in different languages

ElenaKusevska commented:
@rgaudin I think I addressed the above issues. Let me know if it's OK now, or I should make some additional corrections.

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+        prog=NAME,
+        description="Scraper to create ZIM files wikihow articles",
+    )
+    #["en", "es", "pt", "it", "fr", "ru", "de", "zh", "nl", "cz", "id", "jp", "hi", "th", "ar", "vn"],
```
We don't need that comment

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+        default="en",
+        help="The language is specified using the language code. "
+                "The default language is English. "
+                "You may choose from one of the following languages that wikihow "
```
Build that dynamically using the constants. You can either Store languages codes, names and URLs in the same dict or just not care about the language name and just display the codes here. There are not that many and I think those are self-explainatory

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+    parser.add_argument(
+        "--tags",
+        help="List of comma-separated Tags for the ZIM file. "
+        "category:other, kolibri, and _videos:yes added automatically",
```
We certainly won't automatically add `kolibri` but `wikihow` instead

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+
+    parser.add_argument(
+        "--title",
+        help="Custom title for your ZIM. Wikihow article title otherwise",
```
“*Wikihow article title*” doesn't mean anything in this context. I suppose we'll use the homepage title as default

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+
+    parser.add_argument(
+        "--description",
+        help="Custom description for your ZIM. Wikihow article description otherwise",
```
Same as for title. I suppose we'll use the `<meta name="description" />` tag content as default

rgaudin commented on setup.py:
```diff
@@ -0,0 +1,47 @@
+        ]
+    },
+    classifiers=[
+        "Development Status :: ",
```
Remove if not set ; or use the development classifier

rgaudin commented on setup.py:
```diff
@@ -0,0 +1,47 @@
+    },
+    classifiers=[
+        "Development Status :: ",
+        "Intended Audience :: ",
```
Idem

rgaudin commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+requests>=2.26.0,<2.27
+Jinja2>=2.1,<3.0.1
+libzim==1.0.0.dev1
```
We definitely want ranges for libzim and scraperlib as well. I suggest:

```
libzim>=1.0.0.dev1,<1.1
zimscraperlib>=1.4.0.dev1,<1.5
```

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -0,0 +1,36 @@
+        self.url = URLS[self.language]
+
+    def run(self):
+        pass
```
Use the logger to print something so we get a feedback

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+    parser.add_argument(
+        "--language",
+        choices=list(URLS.keys()),
+        default="en",
```
Language should be mandatory

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -0,0 +1,144 @@
+    parser.add_argument(
+        "--name",
+        help="ZIM name. Used as identifier and filename (date will be appended)",
+        required=True,
```
I think we can drop the requirement as the language will allow us to compute a standard Name like`wikihow_{lang}_all`

rgaudin reviewed (changes requested):
Thanks for the fixes 👍
I've added a few minor tweaks to make. `constants.py` and `entrypoint.py` still don't pass black. Make sure you hook it up to your text editor ; probably easier this way.

ElenaKusevska committed:
corrections to entrypoint.py and scraper.py

ElenaKusevska commented:
@rgaudin Check latest changes now, and let me know if it's OK.

Commit:
Minor entrypoint fixes

- restored title and description params
- tweaked/simplified language and tags wording
- removed useless check in scraper. argparse already prevents running with incorrect
language code

rgaudin commented:
Thank you. fixes a couple minor things. Check last commit. Also, whenever you're ready for review, you have to click the *re-request review* button. That triggers a notification and the review UI. Mergin this now 👍
